c++/src/kj/compat/http.c++

```diff
@@ -3627,6 +3959,23 @@ public:
     connectionHeaders[HttpHeaders::BuiltinIndices::SEC_WEBSOCKET_VERSION] = "13";
     connectionHeaders[HttpHeaders::BuiltinIndices::SEC_WEBSOCKET_KEY] = keyBase64;
 
+    kj::Maybe<kj::String> offeredExtensions;
+    kj::Maybe<CompressionParameters> clientOffer;
+    KJ_IF_MAYBE(value, headers.get(HttpHeaderId::SEC_WEBSOCKET_EXTENSIONS)) {
+      // Strip all `Sec-WebSocket-Extensions` except for `permessage-deflate`.
+      kj::StringPtr permittedExtension = "permessage-deflate"_kj;
+      auto extensions = _::splitPartsWithCondition(*value, ',', permittedExtension);
+
+      // Construct our final extensions string.
+      if (extensions.size() > 0) {
+        clientOffer = _::retainClientCompressorConfig(extensions.front());
```

Comment:
The client can offer the same extension multiple times, but offers are ordered by preference. For example, a client might offer `server_max_window_bits`, a parameter the server is not required to support. The offer will be declined if the server doesn't support the parameter, and if no other offer (of lower preference to the client) is provided, then the server won't agree to use any compression. There's an example in the RFC [here](https://datatracker.ietf.org/doc/html/rfc7692#section-7.1.3) (page 18).

I made it only retain the first one since it would be the highest preference option for the client. It wouldn't really make sense for the client to provide different _values_ across offers since values aren't the cause of extensions being declined (unless they're out of range). By values I mean the value 10 in `client_max_window_bits=10`.

IMO, the "most correct" behavior would be to save the config of the first _valid_ offer. Here's some reasons why:
1. The first offer might be invalid, in which case we currently do not retain anything about the client's offer. Retaining the first _valid_ offer fixes this.
2. The `client_x` params are either accepted or ignored. Their presence cannot result in the server outright declining the offer (in the way the `server_x` params might), so there's no reason for a `client_x` param present in offer `n` to be excluded from any lower preference (subsequent) offers `n+k` where `0 < k`.

To be a bit more concrete, let's say the client provided 3 valid offers that include:
1. `client_max_window_bits=14`
2. `client_max_window_bits=10`
3. `client_max_window_bits=15`

respectively. We'll retain the first one, so `clientOffer = 14`.

If the server agrees to offer (1), then all is good and we'll use `14`. If the server agrees to offer (2), then we'll pick `min(clientOffer, serverAgreement) = min(14, 10) = 10`, which is fine. If the server agrees to offer (3), we'll pick `min(14, 15) = 14`, which might seem weird, but is still valid because technically the server "agrees" by responding with a value less than or equal to that which the client offered. If the server doesn't respond with `client_max_window_bits` at all (but still agrees to use compression), we'll just use `clientOffer=14` (which I'd argue is fine since the client gave multiple values, so it doesn't care for any one in particular).

Reply:
Did this in 2 parts since I found a small bug.
Renamed a parameter in `validateCompressionConfig` that caused confusion: https://github.com/capnproto/capnproto/pull/1480/commits/f6c0e5009063f1dd09414e97a0f47e51f6a5e5be
Made `retainClientCompressorConfig` retain the first valid offer + bugfix: https://github.com/capnproto/capnproto/pull/1480/commits/67c479ad0d1c51e1d34df55015774692567f8d0b

Brief description of the bug:
> We were validating the client's offers as though they were in a Response, in other words, as though they were agreed parameters.

This is wrong because a client can make an offer such as `client_max_window_bits; client_no_context_takeover`. If the max bits parameter has no value in an offer, it's just a hint to the server that the client supports it, but this parameter must have a value _in a Response_. Since we were previously validating the offer as though it was an agreement, `validateCompressionParameters` would return `nullptr` it got the offer above, which meant we wouldn't retain `client_no_context_takeover`. The new behavior is to record `client_max_window_bits = 15` and `client_no_context_takeover = true`.

Reply:
But, like, shouldn't we remove all invalid offers from the list before sending them to the server? Otherwise the server might accept an offer that we're not prepared to support?

Reply:
Hmmm, I guess that depends on what you mean by "accept an offer that we're not prepared to support". We already drop every extension that doesn't start with `permessage-deflate`. Assuming that's fine, if the offer is invalid then the server should drop it, but even if the server doesn't follow the spec, our client will parse the response and fail to connect if it finds anything invalid. We could definitely drop invalid offers on the client side before sending them off, it would just require a little more processing on our end.

Reply:
What if the application tries to set some option that is specified by a future version of the spec which we don't support yet, and the server accepts that option? The server will then agree to an offer that we can't actually support, and the connection will fail. That doesn't seem like what we want.

The correct behavior here is to filter out all offers that we don't support.